packages/teleterm/src/main.ts

```diff
@@ -72,10 +73,25 @@ function initializeApp(): void {
     }
   );
 
-  app.on('will-quit', () => {
+  app.on('will-quit', async event => {
+    event.preventDefault();
+
     appStateFileStorage.putAllSync();
     globalShortcut.unregisterAll();
-    mainProcess.dispose();
+    await mainProcess.dispose();
+    app.exit();
+  });
+
+  app.on('quit', () => {
```

Comment:
Why do we need to handle it within a separate callback? Couldn't we still have that in the handler for F6 where we'd first await `mainProcess.dispose()` and then perform the rest of the logic?

Reply:
I don't think it can work, because it would result in waiting for `mainProcese.dispose()` two times:
1. F6 is pressed - we wait for the processes to close
2. They are closed, we spawn a new instance and call `app.quit()`
3. `app.on('will-quit')` is called
4. We wait for `mainProcess.dispose()` the second time (it results in a timeout)
5. _Attempted to open a second instance of the app, exiting._

TBH, it seems to me that my changes have cleaned things up a bit:
1. We don't call `mainProcess.dispose` twice
2. Things are called in the correct order - the new process is spawned after the functions from `will-quit` are called, not before.

Reply:
> TBH, it seems to me that my changes have cleaned things up a bit…

Right, I'd say your changes made things more correct but they also increased complexity a little bit as we now have an extra event handler to deal with and an extra variable that's used in different handlers. 😏

I was thinking that in dev mode, we don't really care about gracefully stopping anything so we might as well kill both child processes.

OTOH it's nice that we have a single handler for cleaning up stuff that's called in both cases so this change sounds fine to me.